datafusion/optimizer/src/rewrite_disjunctive_predicate.rs

```diff
@@ -0,0 +1,429 @@
+                };
+            }
+            and_expr
+        }
+        Predicate::Or { args } => {
+            assert!(args.len() >= 2);
+            let left = normalize_predicate(&args[0]);
+            let right = normalize_predicate(&args[1]);
+            let mut or_expr = BinaryExpr {
+                left: Box::new(left),
+                op: Operator::Or,
+                right: Box::new(right),
+            };
+            for arg in args.iter().skip(2) {
+                or_expr = BinaryExpr {
+                    left: Box::new(or_expr),
+                    op: Operator::Or,
+                    right: Box::new(normalize_predicate(arg)),
+                };
+            }
+            or_expr
+        }
+        Predicate::Other { expr } => *expr.clone(),
+    }
+}
+
+fn rewrite_predicate(predicate: &Predicate) -> Predicate {
+    match predicate {
+        Predicate::And { args } => {
+            let mut rewritten_args = Vec::with_capacity(args.len());
+            for arg in args.iter() {
+                rewritten_args.push(rewrite_predicate(arg));
+            }
+            rewritten_args = flatten_and_predicates(&rewritten_args);
+            Predicate::And {
+                args: rewritten_args,
+            }
+        }
+        Predicate::Or { args } => {
+            let mut rewritten_args = vec![];
+            for arg in args.iter() {
+                rewritten_args.push(rewrite_predicate(arg));
+            }
+            rewritten_args = flatten_or_predicates(&rewritten_args);
+            delete_duplicate_predicates(&rewritten_args)
+        }
+        Predicate::Other { expr } => Predicate::Other {
+            expr: Box::new(*expr.clone()),
+        },
+    }
+}
+
+fn flatten_and_predicates(and_predicates: &[Predicate]) -> Vec<Predicate> {
+    let mut flattened_predicates = vec![];
+    for predicate in and_predicates {
+        match predicate {
+            Predicate::And { args } => {
+                flattened_predicates
+                    .extend_from_slice(flatten_and_predicates(args).as_slice());
+            }
+            _ => {
+                flattened_predicates.push(predicate.clone());
+            }
+        }
+    }
+    flattened_predicates
+}
+
+fn flatten_or_predicates(or_predicates: &[Predicate]) -> Vec<Predicate> {
+    let mut flattened_predicates = vec![];
+    for predicate in or_predicates {
+        match predicate {
+            Predicate::Or { args } => {
+                flattened_predicates
+                    .extend_from_slice(flatten_or_predicates(args).as_slice());
+            }
+            _ => {
+                flattened_predicates.push(predicate.clone());
+            }
+        }
+    }
+    flattened_predicates
+}
+
+fn delete_duplicate_predicates(or_predicates: &[Predicate]) -> Predicate {
+    let mut shortest_exprs: Vec<Predicate> = vec![];
+    let mut shortest_exprs_len = 0;
+    // choose the shortest AND predicate
```

Comment:
> I don't understand the need for checking the shortest AND predicate

The shortest AND predicate could be the common expression to be extracted if each of its elements appears in all OR predicates.

> why not check all elements

We don't need to check all elements, only the shortest could be the common expression.

> keeping a set of expressions

Yes, `shortest_exprs` should be a `HashSet`, but to avoid implementing `Eq` and `Hash`(it'll be spread deeply), I use `Vec` instead.

Reply:
> We don't need to check all elements, only the shortest could be the common expression.

I don't understand why only the shortest could be the common expression. I am probably missing something. 

For example, in the following predicate, the common sub expression`(p_partkey = l_partkey OR p_partkey > 5)` is not the shortest 

```
  (
            (p_partkey = l_partkey OR p_partkey > 5)
            and p_brand = 'Brand#12'
        )
    or
    (
            (p_partkey = l_partkey OR p_partkey > 5)
            and p_size between 1 and 10
        )
    or
    (
             (p_partkey = l_partkey OR p_partkey > 5)
            and p_size between 1 and 15
        )";
```

and yet it could be factored out

```
 (p_partkey = l_partkey OR p_partkey > 5)
and 
  (
            p_brand = 'Brand#12'
        )
    or
    (
            p_size between 1 and 10
        )
    or
    (
            p_size between 1 and 15
        )";
```

🤔

Reply:
That being said, I don't think it is incorrect to pick the shortest predicate, but I do think it may miss potential rewrites. We can always improve it in the future perhaps

Reply:
> For example, in the following predicate, the common sub expression`(p_partkey = l_partkey OR p_partkey > 5)` is not the shortest

The common sub-expression is from the shortest AND predicate, but the shortest AND predicate is not equal to the common sub-expression(beside each element in the shortest AND predicate is in all the OR arguments.)